Plugins/NativeXr/Source/NativeXr.cpp

```diff
@@ -609,68 +623,99 @@ namespace Babylon
-                        bgfx::overrideInternal(colorTexture, reinterpret_cast<uintptr_t>(view.ColorTexturePointer));
-                        bgfx::overrideInternal(depthTexture, reinterpret_cast<uintptr_t>(view.DepthTexturePointer));
-                    }).then(m_runtimeScheduler, m_sessionState->CancellationSource, [this, thisRef{shared_from_this()}, colorTexture, depthTexture, &view]() {
-                        std::array<bgfx::Attachment, 2> attachments{};
-                        attachments[0].init(colorTexture);
-                        attachments[1].init(depthTexture);
-                        auto frameBufferHandle = bgfx::createFrameBuffer(static_cast<uint8_t>(attachments.size()), attachments.data(), false);
-
-                        auto& frameBuffer{m_sessionState->GraphicsImpl.AddFrameBuffer(frameBufferHandle,
-                            static_cast<uint16_t>(view.ColorTextureSize.Width),
-                            static_cast<uint16_t>(view.ColorTextureSize.Height),
-                            true)};
-
-                        // WebXR, at least in its current implementation, specifies an implicit default clear to black.
-                        // https://immersive-web.github.io/webxr/#xrwebgllayer-interface
-                        frameBuffer.Clear(m_sessionState->GraphicsImpl.GetUpdateToken().GetEncoder(), BGFX_CLEAR_COLOR | BGFX_CLEAR_DEPTH | BGFX_CLEAR_STENCIL, 0, 1.0f, 0);
-
-                        m_sessionState->TextureToFrameBufferMap[view.ColorTexturePointer] = &frameBuffer;
-
-                        auto jsWidth{Napi::Value::From(m_env, view.ColorTextureSize.Width)};
-                        auto jsHeight{Napi::Value::From(m_env, view.ColorTextureSize.Height)};
-                        auto jsFrameBuffer{Napi::External<FrameBuffer>::New(m_env, &frameBuffer)};
-                        m_sessionState->FrameBufferToJsTextureMap[&frameBuffer] = Napi::Persistent(m_sessionState->CreateRenderTexture.Call({jsWidth, jsHeight, jsFrameBuffer}).As<Napi::Object>());
+                    auto depthTexture = bgfx::createTexture2D(textureWidth, textureHeight, false, textureLayers, depthTextureFormat, BGFX_TEXTURE_RT);
+                    m_sessionState->GraphicsImpl.AddTexture(depthTexture, textureWidth, textureHeight, false, textureLayers, depthTextureFormat);
+
+                    arcana::make_task(m_sessionState->GraphicsImpl.AfterRenderScheduler(), arcana::cancellation::none(), [colorTexture, depthTexture, &viewConfig]() {
+                        bgfx::overrideInternal(colorTexture, reinterpret_cast<uintptr_t>(viewConfig.ColorTexturePointer));
+                        bgfx::overrideInternal(depthTexture, reinterpret_cast<uintptr_t>(viewConfig.DepthTexturePointer));
+                    }).then(m_runtimeScheduler, m_sessionState->CancellationSource, [this, thisRef{shared_from_this()}, colorTexture, depthTexture, &viewConfig]() {
```

Comment:
>this, thisRef{shared_from_this()} [](http://example.com/codeflow?start=85&length=33)

is there a reason we pass both a this ref and shared_from_this ref? Are we just trying to guarantee that this continues to exist? Would it be better if this was actually a weak reference that we then attempted to resolve when executing the llambda?

Reply:
Seems like @bghgary may have written those lines? I'm not really sure when it comes to this.

Reply:
This is often a common pattern you see people do when trying to ensure a shared pointer lives through the duration of the lambda. You capture both `this` which is itself a non-owning pointer and `shared_from_this()` which extends the lifetime.

If you wanted to, you could omit the `this` and always reference through the `thisRef`, it just adds boilerplate code.